test/hierarchical/test_ordinal.py

```diff
@@ -0,0 +1,310 @@
+            n_threads=1,
+            x_ids=petab_problem.x_ids,
+            parameter_mapping=problem.objective.parameter_mapping,
+            fim_for_hess=False,
+        )
+
+    x_dct = dict(zip(petab_problem.x_ids, petab_problem.x_nominal_scaled))
+
+    calculator_results = {
+        method: calculate(problems[method], x_dct=x_dct) for method in methods
+    }
+
+    # The results of the objective gradient and function value
+    # should not depend on the method given.
+    assert np.isclose(
+        calculator_results[STANDARD]['fval'],
+        calculator_results[REDUCED]['fval'],
+    )
+    assert np.isclose(
+        calculator_results[STANDARD]['grad'],
+        calculator_results[REDUCED]['grad'],
+    ).all()
+
+    # Since the nominal parameters are close to true ones,
+    # the fval and grad should both be low.
+    assert np.all(calculator_results[STANDARD]['fval'] < 0.2)
+    assert np.all(calculator_results[STANDARD]['grad'] < 0.1)
+
+
+def _inner_problem_exp():
+    timepoints = np.linspace(1, 10, 10)
+
+    expected_inner_parameter_values = {
+        STANDARD: np.asarray(
+            [
+                2.07281338e-14,
+                1.99999999e00,
+                2.58823529e00,
+                4.00689557e00,
+                4.59542030e00,
+                6.23751538e00,
+                6.82578436e00,
+                8.40233862e00,
+                8.99060060e00,
+                1.10767572e01,
+            ]
+        ),
+        REDUCED: np.asarray(
+            [1.99999999, 4.00689557, 6.23751538, 8.40233862, 11.0767572]
+        ),
+    }
+
+    simulation = timepoints
+    data = np.full(simulation.shape, np.nan)
+
+    par_types = ['cat_lb', 'cat_ub']
+
+    categories = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
+    inner_parameter_ids = [
+        f'{par_type}_{category}'
+        for par_type in par_types
+        for category in set(categories)
+    ]
+
+    # Get data masks for inner parameters
+    masks = []
+    for category in categories:
+        new_mask = np.full(data.shape, False)
+        new_mask[2 * (category - 1) : 2 * category] = True
+        masks.append([new_mask])
+
+    # Construct inner parameters
+    inner_parameters = [
+        OptimalScalingParameter(
+            inner_parameter_id=inner_parameter_id,
+            inner_parameter_type=InnerParameterType.OPTIMAL_SCALING,
+            scale=LIN,
+            lb=-np.inf,
+            ub=np.inf,
+            ixs=mask,
+            group=1,
+            category=inner_parameter_category,
+        )
+        for inner_parameter_id, mask, inner_parameter_category in zip(
+            inner_parameter_ids, masks, categories
+        )
+    ]
+
+    # Construct inner problem
+    inner_problem = OptimalScalingProblem(
+        xs=inner_parameters, data=[data], method=STANDARD
+    )
+
+    return inner_problem, expected_inner_parameter_values, simulation
+
+
+def test_optimal_scaling_solver():
+    """Test the Optimal scaling solver."""
```

Comment:
what exactly does it test?

Reply:
It tests the inner solver directly using a very simple synthetic example from `_inner_problem_exp()`. 
Since the problem is so simple, it should solve it with fval=0 and grad=0 and get the `expected_inner_parameter_values`.